cuqi/utilities/_utilities.py

```diff
@@ -173,18 +163,46 @@ def approx_derivative(func, wrt, direction=None, epsilon=np.sqrt(np.finfo(np.flo
     # corresponding geometry par2fun map is not identity
     # (e.g. funvalues=paramters**2), because wrt entries 
     # are interpreted as function value.
-    if isinstance(wrt, CUQIarray) or isinstance(direction, CUQIarray):
-        raise NotImplementedError("approx_derivative is not implemented"+
-                                  "for inputs of type CUQIarray")
+
+    if isinstance(wrt, CUQIarray):
```

Comment:
Please add more comments explaining the following ca 20 lines.

Reply:
I added more documentation